src/cdk/dialog/dialog-config.ts

```diff
@@ -57,6 +57,13 @@ export class DialogConfig<D = unknown, R = unknown, C extends BasePortalOutlet =
   /** Whether the dialog closes with the escape key or pointer events outside the panel element. */
   disableClose?: boolean = false;
 
+  /** Function used to determine whether the dialog is allowed to close. */
+  closePredicate?: <Result = unknown, Component = unknown>(
```

Comment:
I'm somewhat on the fence about this API, because technically you can do the same by combining existing APIs and being careful with your event handlers, but I decided to implement it since there's clearly demand for it in #14292. 

There's an overlap with `disableClose` and initially my thought was to change `disableClose` to also accept a function, but the problem is that it only applies to attempts at closing the dialog using the backdrop or the escape key, not programmatic closes or by something like the `matDialogClose` directive.

Reply:
The idea with the new API is that it blocks all the `close` calls, even the programmatic ones, whereas `disableClose` was only for the backdrop and escape keys. The primary use case for this is if you have a form and you want to prevent the user from accidentally closing it based on a certain condition.

Reply:
I can be on board with this if we rename `disableClose` to avoid confusion between the two (deprecating the current name for the appropriate time)